src/bigint.rs

```diff
@@ -116,6 +116,54 @@ pub mod big_digit {
     }
 }
 
+/*
+ * Generic functions for add/subtract/multiply with carry/borrow:
+ */
+
+// Add with carry:
+#[inline]
+fn adc(a: BigDigit, b: BigDigit, carry: &mut BigDigit) -> BigDigit {
+    let (hi, lo) = big_digit::from_doublebigdigit(
+        (a as DoubleBigDigit) +
+        (b as DoubleBigDigit) +
+        (*carry as DoubleBigDigit));
+
+    *carry = hi;
+    lo
+}
+
+// Subtract with borrow:
+#[inline]
+fn sbb(a: BigDigit, b: BigDigit, borrow: &mut BigDigit) -> BigDigit {
+    let (hi, lo) = big_digit::from_doublebigdigit(
+        big_digit::BASE
+        + (a as DoubleBigDigit)
+        - (b as DoubleBigDigit)
+        - (*borrow as DoubleBigDigit));
+    /*
+       hi * (base) + lo == 1*(base) + ai - bi - borrow
+       => ai - bi - borrow < 0 <=> hi == 0
+       */
+    *borrow = if hi == 0 { 1 } else { 0 };
+    lo
+}
+
+#[inline]
+fn mac_with_carry(a: BigDigit, b: BigDigit, c: BigDigit, carry: &mut BigDigit) -> BigDigit {
+    let (hi, lo) = big_digit::from_doublebigdigit(
+        (a as DoubleBigDigit) +
+        (b as DoubleBigDigit) * (c as DoubleBigDigit) +
+        (*carry as DoubleBigDigit));
+    *carry = hi;
+    lo
+}
+
+#[inline]
+fn div_wide(hi: BigDigit, lo: BigDigit, divisor: BigDigit) -> (BigDigit, BigDigit) {
+    let lhs = big_digit::to_doublebigdigit(hi, lo);
+    let rhs = divisor as DoubleBigDigit;
+    ((lhs / rhs) as BigDigit, (lhs % rhs) as BigDigit)
+}
```

Comment:
On Mon, Nov 30, 2015 at 03:39:08PM -0800, Josh Stone wrote:
> > +#[inline]
> > +fn mac_with_carry(a: BigDigit, b: BigDigit, c: BigDigit, carry: &mut BigDigit) -> BigDigit {
> > +    let (hi, lo) = big_digit::from_doublebigdigit(
> > +        (a as DoubleBigDigit) +
> > +        (b as DoubleBigDigit) * (c as DoubleBigDigit) +
> > +        (*carry as DoubleBigDigit));
> > +    *carry = hi;
> > +    lo
> > +}
> > +
> > +#[inline]
> > +fn div_wide(hi: BigDigit, lo: BigDigit, divisor: BigDigit) -> (BigDigit, BigDigit) {
> > +    let lhs = big_digit::to_doublebigdigit(hi, lo);
> > +    let rhs = divisor as DoubleBigDigit;
> > +    ((lhs / rhs) as BigDigit, (lhs % rhs) as BigDigit)
> > +}
> 
> Considered alone, this looks like a bad assumption that `lhs/rhs` will always fit into `BigDigit`.  But since this is only called from `div_rem_digit` with prior remainders in `hi`, it should be fine.  Perhaps this deserves a comment and `debug_assert!(hi < divisor)`, or just fold this code into `div_rem_digit` where it's known to be safe.

Like I mentioned, this is implementing exactly what the x86 div instruction does
- I'll add a comment explaining the caller has to know the result will fit into
a BigDigit and that that's not true in general.